### PR TITLE
Query Gson adapter once to avoid invoking gson.fromJson inside loops

### PR DESCRIPTION
**Pull request by Egorand**

### CODE REVIEW

File: wire-gson-support/src/main/java/com/squareup/wire/MessageTypeAdapter.kt [hunk 1]
```diff
@@ -124,11 +125,12 @@ internal class MessageTypeAdapter<M : Message<M, B>, B : Message.Builder<M, B>>(
 
       val keyType = fieldBinding.keyAdapter().javaType
       val valueType = fieldBinding.singleAdapter().javaType
+      val valueAdapter = gson.getAdapter(valueType)
 
       val jsonObject = element.asJsonObject
       return jsonObject.entrySet().associateBy(
           { gson.fromJson(it.key, keyType) },
```
Comment (Egorand) [inline wire-gson-support/src/main/java/com/squareup/wire/MessageTypeAdapter.kt]: Tried to do the same for the key, but its type is inferred as String, and `keyAdapter.fromJson(String)` fails tests. Is there a correct way to do this?

Comment (JakeWharton) [reply wire-gson-support/src/main/java/com/squareup/wire/MessageTypeAdapter.kt]: What's the failure look like?

Comment (Egorand) [reply wire-gson-support/src/main/java/com/squareup/wire/MessageTypeAdapter.kt]: `com.google.gson.JsonSyntaxException: com.google.gson.stream.MalformedJsonException: Use JsonReader.setLenient(true) to accept malformed JSON at line 1 column 1 path $`

Comment (JakeWharton) [reply wire-gson-support/src/main/java/com/squareup/wire/MessageTypeAdapter.kt]: Gson might not support top-level strings despite being valid JSON. I can't remember. You can do what it says and create your own `JsonReader` and set lenient to true, but probably not worth the effort.

Comment (Egorand) [reply wire-gson-support/src/main/java/com/squareup/wire/MessageTypeAdapter.kt]: Ah, makes sense. Will leave as is for now.

**Pull request merged**